runtime/compiler/control/CompilationThread.cpp

```diff
@@ -2901,6 +2899,76 @@ bool TR::CompilationInfo::suspendCompThreadsForCheckpoint(J9VMThread *vmThread)
    return true;
    }
 
+bool
+TR::CompilationInfo::suspendCompilerThreadsForCheckpoint(J9VMThread *vmThread)
+   {
+   // Suspend compilation threads for checkpoint
+   if (!suspendCompThreadsForCheckpoint(vmThread))
+      return false;
+
+   // Suspend Sampler Thread
+   if (_jitConfig->samplerMonitor)
+      {
+      j9thread_monitor_enter(_jitConfig->samplerMonitor);
+      j9thread_interrupt(_jitConfig->samplerThread);
+
+      // Determine whether to wait on the CR Monitor.
+      //
+      // Note, this thread releases the sampler monitor and then
+      // acquires the CR monitor inside releaseCompMonitorUntilNotifiedOnCRMonitor.
+      while (!shouldCheckpointBeInterrupted()
+             && getSamplingThreadLifetimeState() != TR::CompilationInfo::SAMPLE_THR_SUSPENDED)
+         {
+         j9thread_monitor_exit(_jitConfig->samplerMonitor);
+         releaseCompMonitorUntilNotifiedOnCRMonitor(vmThread);
+         j9thread_monitor_enter(_jitConfig->samplerMonitor);
+         }
+
+      j9thread_monitor_exit(_jitConfig->samplerMonitor);
+      }
+
+   return !shouldCheckpointBeInterrupted();
+   }
+
+void
+TR::CompilationInfo::resumeCompilerThreadsForRestore(J9VMThread *vmThread)
+   {
+   // Resume suspended Sampler Thread
+   if (_jitConfig->samplerMonitor)
+      {
+      j9thread_monitor_enter(_jitConfig->samplerMonitor);
+      j9thread_monitor_notify_all(_jitConfig->samplerMonitor);
+      j9thread_monitor_exit(_jitConfig->samplerMonitor);
+      }
+
+   // Resume suspended compilation threads.
+   resumeCompilationThread();
+   }
+
+void
+TR::CompilationInfo::resetStartAndElapsedTime()
+   {
+   // Post-restore, reset the start and elapsed time. The Checkpoint
+   // phase is conceptually part of building the application; therefore
+   // it does not make sense to expect a user who specifies an option such
+   // as -XsamplingExpirationTime  to take into account the time spent
+   // executing in the Checkpoint phase.
+
+   PORT_ACCESS_FROM_JAVAVM(jitConfig->javaVM);
+   TR::PersistentInfo *persistentInfo = getPersistentInfo();
+
+   if (TR::Options::isAnyVerboseOptionSet())
+      TR_VerboseLog::writeLineLocked(TR_Vlog_CHECKPOINT_RESTORE, "Start and elapsed time: startTime=%6u, elapsedTime=%6u",
+                                       (uint32_t)persistentInfo->getStartTime(), (uint32_t)persistentInfo->getElapsedTime());
+
+   persistentInfo->setStartTime(j9time_current_time_millis());
+   persistentInfo->setElapsedTime(0);
```

Comment:
I suppose if we want to keep the Elapsed Time the same, we could set the start time to be `j9time_current_time_millis() - elapsedtime`.